ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/api/BatchingLedgerWriterSpec.scala

```diff
@@ -0,0 +1,159 @@
+import com.daml.ledger.participant.state.kvutils.Envelope
+import com.daml.ledger.participant.state.v1
+import com.daml.ledger.participant.state.v1.SubmissionResult
+import com.digitalasset.ledger.api.testing.utils.AkkaBeforeAndAfterAll
+import com.digitalasset.logging.LoggingContext
+import com.google.protobuf.ByteString
+import org.mockito.{ArgumentCaptor, ArgumentMatchers}
+import org.mockito.Mockito.{timeout, times, _}
+import org.scalatest.mockito.MockitoSugar
+import org.scalatest.{AsyncWordSpec, Matchers}
+
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+class BatchingLedgerWriterSpec
+    extends AsyncWordSpec
+    with MockitoSugar
+    with AkkaBeforeAndAfterAll
+    with Matchers {
+
+  private def createWriter(
+      captor: Option[ArgumentCaptor[Array[Byte]]] = None,
+      result: SubmissionResult = SubmissionResult.Acknowledged): LedgerWriter = {
+    val writer = mock[LedgerWriter]
+    when(
+      writer.commit(
+        ArgumentMatchers.anyString(),
+        captor.map(_.capture()).getOrElse(ArgumentMatchers.any[Array[Byte]]())))
+      .thenReturn(Future.successful(SubmissionResult.Acknowledged))
+    when(writer.participantId).thenReturn(v1.ParticipantId.assertFromString("test-participant"))
+    writer
+  }
+
+  private def createBatchingWriter(
+      writer: LedgerWriter,
+      maxBatchSizeBytes: Long,
+      maxWaitDuration: FiniteDuration) =
+    LoggingContext.newLoggingContext { implicit logCtx =>
+      new BatchingLedgerWriter(
+        writer = writer,
+        maxQueueSize = 128,
+        maxBatchSizeBytes = maxBatchSizeBytes,
+        maxWaitDuration = maxWaitDuration,
+        maxParallelism = 1
+      )
+    }
+
+  private def createExpectedBatch(corId: String, subm: Array[Byte]) =
+    Envelope
+      .enclose(
+        DamlSubmissionBatch.newBuilder
+          .addSubmissions(
+            DamlSubmissionBatch.CorrelatedSubmission.newBuilder
+              .setCorrelationId(corId)
+              .setSubmission(ByteString.copyFrom(subm)))
+          .build)
+      .toByteArray
+
+  "BatchingLedgerWriter" should {
+
+    // Test that even when maxWaitDuration is reached we don't commit an empty batch.
+    "not commit empty batches" in {
+      val mockWriter = mock[LedgerWriter]
+      val batchingWriter = createBatchingWriter(mockWriter, 1024, 10.millis)
+      Thread.sleep(2 * batchingWriter.maxWaitDuration.toMillis);
+      verifyZeroInteractions(mockWriter)
+      succeed
+    }
+
+    // Test that we commit a batch when we have a submission and the maxWaitDuration has been reached.
+    "commit batch after maxWaitDuration" in {
+      val batchCaptor: ArgumentCaptor[Array[Byte]] = ArgumentCaptor.forClass(classOf[Array[Byte]])
+      val mockWriter = createWriter(Some(batchCaptor))
+      val batchingWriter = createBatchingWriter(
+        mockWriter,
+        1024,
+        /* Fingers crossed that this is long enough to not make this test flaky */
```

Comment:
We'll soon know :) I don't think there's a perfect way to run time-sensitive tests but perhaps we should start considering something like https://github.com/puniverse/timewarp